{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "import zipfile\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "def parse_annotation_file(annotation_fn):\n",
    "    \"\"\"\n",
    "    :param annotation_fn: the gene annotations file\n",
    "    :return: outputs a list of tuples, \"genes\". Every tuple represents one gene, the first element of the tuple is the\n",
    "            list of exon index ranges for that gene, the second element of the tuple is the list of isoforms that exist\n",
    "            for that gene. For example, genes[0][0][0] references the first exon range of the first gene (in a tuple),\n",
    "            genes[2][1][3] references the fourth isoform of the third gene.\n",
    "    \"\"\"\n",
    "\n",
    "    with open(annotation_fn, 'r') as aFile:\n",
    "        N = int(aFile.readline().strip())\n",
    "        genes = [None]*N\n",
    "        for i in range(N):\n",
    "            numExons = int(aFile.readline().strip())\n",
    "            exons = [None]*numExons\n",
    "            starts = [int(x) for x in aFile.readline().strip().split(' ')]\n",
    "            ends = [int(x) for x in aFile.readline().strip().split(' ')]\n",
    "            for j in range(numExons):\n",
    "                exons[j] = (starts[j], ends[j])\n",
    "            numIsoforms = int(aFile.readline().strip())\n",
    "            isoforms = [None]*numIsoforms\n",
    "            for j in range(numIsoforms):\n",
    "                isoforms[j] = [int(x) for x in aFile.readline().strip().split(' ')]\n",
    "            genes[i] = (exons, isoforms)\n",
    "    return genes\n",
    "\n",
    "\n",
    "def parse_genome_file(genome_fn):\n",
    "    \"\"\"\n",
    "    :param genome_fn: the full genome file\n",
    "    :return: the string containing the genome\n",
    "    \"\"\"\n",
    "\n",
    "    with open(genome_fn, 'r') as gFile:\n",
    "        return gFile.readline().strip()\n",
    "\n",
    "\n",
    "def parse_reads_file(reads_fn):\n",
    "    \"\"\"\n",
    "    :param reads_fn: the file of shuffled reads\n",
    "    :return: a list containing all of the shuffled reads\n",
    "    \"\"\"\n",
    "    \n",
    "    out_reads = []\n",
    "    with open(reads_fn, 'r') as rFile:\n",
    "        for line in rFile:\n",
    "            out_reads.append(line.strip())\n",
    "    return out_reads\n",
    "\n",
    "\n",
    "def quantify_isoforms(genes, genome, reads):\n",
    "    ret=[]\n",
    "    for gene in genes:\n",
    "        exons_seq=[]\n",
    "        exons=gene[0]\n",
    "        for exon in exons:\n",
    "            exons_seq=exons_seq+[genome[exon[0]:exon[1]+1]]\n",
    "        isoforms_seq=[]\n",
    "        isoforms=gene[1]\n",
    "        for isoform in isoforms:\n",
    "            iso_seq=''\n",
    "            for iso in isoform:\n",
    "                iso_seq=iso_seq+exons_seq[iso]\n",
    "            isoforms_seq=isoforms_seq+[iso_seq]\n",
    "        isonum=len(isoforms_seq)\n",
    "        freq=1/isonum\n",
    "        for seq in isoforms_seq:\n",
    "            ret=ret+[(seq,freq)]\n",
    "    return ret\n",
    "    \n",
    "    \"\"\"\n",
    "    :param genes: the list of gene tuples generated by the parser\n",
    "    :param genome_fn: the full genome file\n",
    "    :param reads_fn: the file of shuffled reads\n",
    "    :return: a list of tuples, where the first element of the tuple is the transcript sequence (the isoform in terms of\n",
    "            the exon sequences that form it in the genome), and the second element of the tuple is the abundance of that\n",
    "            specific isoform\n",
    "            NOTE: this skeleton is built assuming the return value exists like this, but as long as you change the way\n",
    "            the output file is generated, this can be in whatever form you like.\n",
    "    \"\"\"\n",
    "\n",
    "    \"\"\"\n",
    "        TODO: Within this function, you should go through most of the process of quantifying isoforms given the data.\n",
    "            This can be broken down into the following few steps:\n",
    "        \n",
    "            1. Align reads to the genome, exome, or isoforms\n",
    "                    your choice of method, but note the length of the genome\n",
    "            \n",
    "            2. Use the generated alignment to get exon counts\n",
    "            \n",
    "            3. Formulate your RNA seq problem using the isoforms and exon counts (linear algebra)\n",
    "            \n",
    "            4. Compute the isoform abundances based on your above formulation\n",
    "    \"\"\"\n",
    "\n",
    "genes = parse_annotation_file('DATA_PA_1100_0.txt')\n",
    "genome = parse_genome_file('full_genome.txt')\n",
    "reads = parse_reads_file('shuffled_reads.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(1353310, 1353377),\n",
       " (1353616, 1353678),\n",
       " (1353869, 1353931),\n",
       " (1354069, 1354122)]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes[3][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'caagctagacctttctaggctcaaggacaaggctcgggtagcccagcttgtcgagttgtattgtccag': 767,\n",
       " 'tcgctaggacgcgtggggatgatatagcgcataccgaaggggtggtcttggtctggggccttt': 570,\n",
       " 'tgtatgactacatgccttggggatggtgttgttccactgctagatcgatgttccgcagtcggg': 883,\n",
       " 'ccaatctttcactccgattccggtccgtcgattgaaccactcgcctacaggcga': 914}"
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#gene 4\n",
    "exons_four=[]\n",
    "exon_inds=genes[3][0]\n",
    "for inds in exon_inds:\n",
    "    exons_four=exons_four+[genome[inds[0]:inds[1]+1]]\n",
    "dict_exons={}\n",
    "for ex in exons_four:\n",
    "    dict_exons[ex]=0\n",
    "zero_one=exons_four[0]+exons_four[1]\n",
    "zero_three=exons_four[0]+exons_four[3]\n",
    "one_two=exons_four[1]+exons_four[2]\n",
    "two_three=exons_four[2]+exons_four[3]\n",
    "for r in reads:\n",
    "    if (r in zero_one or r in zero_three) and (r not in exons_four[1]) and (r not in exons_four[3]):\n",
    "        dict_exons[exons_four[0]]+=1\n",
    "    if (r in zero_one or r in one_two) and (r not in exons_four[0]) and (r not in exons_four[2]):\n",
    "        dict_exons[exons_four[1]]+=1\n",
    "    if (r in one_two or r in two_three) and (r not in exons_four[1]) and (r not in exons_four[3]):\n",
    "        dict_exons[exons_four[2]]+=1\n",
    "    if (r in zero_three or r in two_three) and (r not in exons_four[0]) and (r not in exons_four[2]):\n",
    "        dict_exons[exons_four[3]]+=1\n",
    "dict_exons\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[767, 570, 883, 914]"
      ]
     },
     "execution_count": 180,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "x = np.array([[68/50, 68/50, 0], [63/50, 0, 63/50], [0, 0, 63/50],[0,54/50,54/50]])\n",
    "x_t=x.transpose()\n",
    "y=[767,570,883,914]\n",
    "y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ 28.70529086 455.97742978 516.04832461]\n",
      "1000.7310452458482\n"
     ]
    }
   ],
   "source": [
    "x_t_x=np.dot(x_t,x)\n",
    "x_t_x_minues = np.linalg.inv(x_t_x) \n",
    "x_then=np.dot(x_t_x_minues,x_t)\n",
    "print(np.dot(x_then,y))\n",
    "print(sum(np.dot(x_then,y)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[3.4372, 1.8496, 1.5876],\n",
       "       [1.8496, 3.016 , 1.1664],\n",
       "       [1.5876, 1.1664, 4.3416]])"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_t_x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0.4183322177734516, 0.33005104759402387, 0.251616734595531]"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[74.68324102/178.5261518166043, 58.92274343/178.5261518166043,44.92016736/178.5261518166043]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('cccgtcagtgtgtgagcgtgtactcgctatatttgtagtgtgcactgcagaccaagtcggatgtttccttcccgaatctgtgtacacaaaacgtgagatctctcgatgtcacctggcaggtctgggtaagcgtctatgacttttggtgagttgagtccctacgatttgtttctattaacgctgttccatactgacaatgctggtacctaaccgctatccgggttattgtccgtgaggaatgagcagtattcccaaacgtgtcctcacgtgctgtg',\n",
       "  0.5),\n",
       " ('gtacacaaaacgtgagatctctcgatgtcacctggcaggtctgggtaagcgtctatgacttttggtgagttgagtccctacgatttgtttctattaacgctgttccatactgacaatgctggtacctaaccgctatccgggttattgtccgtgaggaatgagcagtattcccaaacgtgtcctcacgtgctgtg',\n",
       "  0.5),\n",
       " ('ccagtagggacgtatttaccggtggaagctggccgagattgtgcgtcacgaacgataaaaatggacaccccgtcgatgtaaggctgtcaaatttgggcgatcggttgctaacaacgcggcatactatatgctttgtgtatgttacgaagtagaggacctcatccaagattgccgccccaacgagtatttcgggaactcctggcaggccggttcgctcgcatatgtactacggccatgttacgggtat',\n",
       "  1.0),\n",
       " ('atcgggcgttaatagatatccaatgatagaccgacggccggggcactgatgtagcaagccgagatgggaatcgctgtatagtgtcaactgtatggaaagtgagtcggggtcagtttgatttatatgagtacagcatttcttgctcgtacctgct',\n",
       "  0.5),\n",
       " ('atcgggcgttaatagatatccaatgatagaccgacggccggggcactgatgtagcaagccgagatgggaatcgctgtatagtgtcaactgtatggaaagtgagtcggggtcagtttgatttatatgagtacagcatttcttgctcgtacctgcttaggtgggttagccctaggttatcatttataatttagtgctctccgatctcgtttgtttagaatgacctatgctagca',\n",
       "  0.5),\n",
       " ('caagctagacctttctaggctcaaggacaaggctcgggtagcccagcttgtcgagttgtattgtccagtcgctaggacgcgtggggatgatatagcgcataccgaaggggtggtcttggtctggggccttt',\n",
       "  0.3333333333333333),\n",
       " ('caagctagacctttctaggctcaaggacaaggctcgggtagcccagcttgtcgagttgtattgtccagccaatctttcactccgattccggtccgtcgattgaaccactcgcctacaggcga',\n",
       "  0.3333333333333333),\n",
       " ('tcgctaggacgcgtggggatgatatagcgcataccgaaggggtggtcttggtctggggccttttgtatgactacatgccttggggatggtgttgttccactgctagatcgatgttccgcagtcgggccaatctttcactccgattccggtccgtcgattgaaccactcgcctacaggcga',\n",
       "  0.3333333333333333),\n",
       " ('acgagcaatatcacctgaccgtttgacggggtacgcaaggtcgtgtctagttggaggctaagcgttacaacgttagcatggtctgatctaggagcctattccgccagaggtgcgtgcgactcattgcgacgaaacctcaattaaggttattgaattttggacatcagaaattgaaggtataggggca',\n",
       "  1.0)]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "quantify_isoforms(genes, genome, reads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "exons_seq=[]\n",
    "for gene in genes:\n",
    "    exons=gene[0]\n",
    "    for exon in exons:\n",
    "        exons_seq=exons_seq+[genome[exon[0]:exon[1]]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1999800"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(reads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "import zipfile\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "def parse_annotation_file(annotation_fn):\n",
    "    \"\"\"\n",
    "    :param annotation_fn: the gene annotations file\n",
    "    :return: outputs a list of tuples, \"genes\". Every tuple represents one gene, the first element of the tuple is the\n",
    "            list of exon index ranges for that gene, the second element of the tuple is the list of isoforms that exist\n",
    "            for that gene. For example, genes[0][0][0] references the first exon range of the first gene (in a tuple),\n",
    "            genes[2][1][3] references the fourth isoform of the third gene.\n",
    "    \"\"\"\n",
    "\n",
    "    with open(annotation_fn, 'r') as aFile:\n",
    "        N = int(aFile.readline().strip())\n",
    "        genes = [None]*N\n",
    "        for i in range(N):\n",
    "            numExons = int(aFile.readline().strip())\n",
    "            exons = [None]*numExons\n",
    "            starts = [int(x) for x in aFile.readline().strip().split(' ')]\n",
    "            ends = [int(x) for x in aFile.readline().strip().split(' ')]\n",
    "            for j in range(numExons):\n",
    "                exons[j] = (starts[j], ends[j])\n",
    "            numIsoforms = int(aFile.readline().strip())\n",
    "            isoforms = [None]*numIsoforms\n",
    "            for j in range(numIsoforms):\n",
    "                isoforms[j] = [int(x) for x in aFile.readline().strip().split(' ')]\n",
    "            genes[i] = (exons, isoforms)\n",
    "    return genes\n",
    "\n",
    "\n",
    "def parse_genome_file(genome_fn):\n",
    "    \"\"\"\n",
    "    :param genome_fn: the full genome file\n",
    "    :return: the string containing the genome\n",
    "    \"\"\"\n",
    "\n",
    "    with open(genome_fn, 'r') as gFile:\n",
    "        return gFile.readline().strip()\n",
    "\n",
    "\n",
    "def parse_reads_file(reads_fn):\n",
    "    \"\"\"\n",
    "    :param reads_fn: the file of shuffled reads\n",
    "    :return: a list containing all of the shuffled reads\n",
    "    \"\"\"\n",
    "    \n",
    "    out_reads = []\n",
    "    with open(reads_fn, 'r') as rFile:\n",
    "        for line in rFile:\n",
    "            out_reads.append(line.strip())\n",
    "    return out_reads\n",
    "\n",
    "\n",
    "def quantify_isoforms(genes, genome, reads):\n",
    "    ret=[]\n",
    "    for gene in genes:\n",
    "        exons_seq=[]\n",
    "        exons=gene[0]\n",
    "        for exon in exons:\n",
    "            exons_seq=exons_seq+[genome[exon[0]:exon[1]+1]]\n",
    "        isoforms_seq=[]\n",
    "        isoforms=gene[1]\n",
    "        for isoform in isoforms:\n",
    "            iso_seq=''\n",
    "            for iso in isoform:\n",
    "                iso_seq=iso_seq+exons_seq[iso]\n",
    "            isoforms_seq=isoforms_seq+[iso_seq]\n",
    "        isonum=len(isoforms_seq)\n",
    "        if len(isoforms_seq)==3:\n",
    "            dict_exons={}\n",
    "            for ex in exons_seq:\n",
    "                dict_exons[ex]=0\n",
    "            zero_one=exons_seq[0]+exons_seq[1]\n",
    "            zero_three=exons_seq[0]+exons_seq[3]\n",
    "            one_two=exons_seq[1]+exons_seq[2]\n",
    "            two_three=exons_seq[2]+exons_seq[3]\n",
    "            for r in reads:\n",
    "                if (r in zero_one or r in zero_three) and (r not in exons_seq[1]) and (r not in exons_seq[3]):\n",
    "                    dict_exons[exons_seq[0]]+=1\n",
    "                if (r in zero_one or r in one_two) and (r not in exons_seq[0]) and (r not in exons_seq[2]):\n",
    "                    dict_exons[exons_seq[1]]+=1\n",
    "                if (r in one_two or r in two_three) and (r not in exons_seq[1]) and (r not in exons_seq[3]):\n",
    "                    dict_exons[exons_seq[2]]+=1\n",
    "                if (r in zero_three or r in two_three) and (r not in exons_seq[0]) and (r not in exons_seq[2]):\n",
    "                    dict_exons[exons_seq[3]]+=1\n",
    "            print(dict_exons)\n",
    "            x = np.array([[68/50, 68/50, 0], [63/50, 0, 63/50], [0, 0, 63/50],[0,54/50,54/50]])\n",
    "            x_t=x.transpose()\n",
    "            y=[dict_exons[exons_seq[0]],dict_exons[exons_seq[1]],dict_exons[exons_seq[2]],dict_exons[exons_seq[3]]]\n",
    "            x_t_x=np.dot(x_t,x)\n",
    "            x_t_x_minues = np.linalg.inv(x_t_x) \n",
    "            x_then=np.dot(x_t_x_minues,x_t)\n",
    "            beta=np.dot(x_then,y)\n",
    "            freq_li=[beta[0]/sum(beta),beta[1]/sum(beta),beta[2]/sum(beta)]\n",
    "            ret=ret+[(isoforms_seq[0],freq_li[0])]\n",
    "            ret=ret+[(isoforms_seq[1],freq_li[1])]\n",
    "            ret=ret+[(isoforms_seq[2],freq_li[2])]\n",
    "        else:\n",
    "            freq=1/isonum\n",
    "            for seq in isoforms_seq:\n",
    "                ret=ret+[(seq,freq)]\n",
    "    return ret\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    \"\"\"\n",
    "    For an example of how you might call this script to run on the data provided:\n",
    "    \n",
    "    Usage: python proj4.py -g full_genome.txt -r shuffled_reads.txt -a DATA_PA_1100_0 -o test.out -t hw4_r_4_chr_1\n",
    "    \"\"\"\n",
    "    parser = argparse.ArgumentParser(description='For now this starter code helps parse the files given, but leaves\\n'\n",
    "                                                 'the actual function that must be implemented empty')\n",
    "    parser.add_argument('-g', '--genome', required=True, dest='genome_file', help='File containing the full genome')\n",
    "    parser.add_argument('-r', '--reads', required=True, dest='read_file', help='File containing the shuffled reads')\n",
    "    parser.add_argument('-a', '--annotation', required=True, dest='annotation_file', help='File containing gene '\n",
    "                                                                                          'annotations')\n",
    "    parser.add_argument('-o', '--outputFile', required=True, dest='output_file', help='Output file name')\n",
    "    parser.add_argument('-t', '--outputHeader', required=True, dest='output_header',\n",
    "                        help='String that needs to be output on the first line of the output file so that the online\\n'\n",
    "                             'submission system recognizes which leaderboard this file should be submitted to. For\\n'\n",
    "                             'hw4, this will be hw4_r_4_chr_1')\n",
    "\n",
    "    args = parser.parse_args()\n",
    "    genome_fn = args.genome_file\n",
    "    reads_fn = args.read_file\n",
    "    annotation_fn = args.annotation_file\n",
    "    output_fn = args.output_file\n",
    "\n",
    "    genes = parse_annotation_file(annotation_fn)\n",
    "    genome = parse_genome_file(genome_fn)\n",
    "    reads = parse_reads_file(reads_fn)\n",
    "    print(len(reads))\n",
    "    output = quantify_isoforms(genes, genome, reads)\n",
    "    with open(output_fn, 'w') as oFile:\n",
    "        oFile.write('>' + args.output_header + '\\n')\n",
    "        oFile.write('>RNA\\n')\n",
    "        for isoform in output:\n",
    "            out_str = '{} {}\\n'.format(isoform[0], isoform[1])\n",
    "            oFile.write(out_str)\n",
    "\n",
    "    zip_fn = output_fn + '.zip'\n",
    "    with zipfile.ZipFile(zip_fn, 'w') as zFile:\n",
    "        zFile.write(output_fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [],
   "source": [
    "def quantify_isoforms(genes, genome, reads):\n",
    "    ret=[]\n",
    "    for gene in genes:\n",
    "        exons_seq=[]\n",
    "        exons=gene[0]\n",
    "        for exon in exons:\n",
    "            exons_seq=exons_seq+[genome[exon[0]:exon[1]+1]]\n",
    "        isoforms_seq=[]\n",
    "        isoforms=gene[1]\n",
    "        for isoform in isoforms:\n",
    "            iso_seq=''\n",
    "            for iso in isoform:\n",
    "                iso_seq=iso_seq+exons_seq[iso]\n",
    "            isoforms_seq=isoforms_seq+[iso_seq]\n",
    "        isonum=len(isoforms_seq)\n",
    "        if len(isoforms_seq)==3:\n",
    "            dict_exons={}\n",
    "            for ex in exons_seq:\n",
    "                dict_exons[ex]=0\n",
    "            zero_one=exons_seq[0]+exons_seq[1]\n",
    "            zero_three=exons_seq[0]+exons_seq[3]\n",
    "            one_two=exons_seq[1]+exons_seq[2]\n",
    "            two_three=exons_seq[2]+exons_seq[3]\n",
    "            for r in reads:\n",
    "                if (r in zero_one or r in zero_three) and (r not in exons_seq[1]) and (r not in exons_seq[3]):\n",
    "                    dict_exons[exons_seq[0]]+=1\n",
    "                if (r in zero_one or r in one_two) and (r not in exons_seq[0]) and (r not in exons_seq[2]):\n",
    "                    dict_exons[exons_seq[1]]+=1\n",
    "                if (r in one_two or r in two_three) and (r not in exons_seq[1]) and (r not in exons_seq[3]):\n",
    "                    dict_exons[exons_seq[2]]+=1\n",
    "                if (r in zero_three or r in two_three) and (r not in exons_seq[0]) and (r not in exons_seq[2]):\n",
    "                    dict_exons[exons_seq[3]]+=1\n",
    "            print(dict_exons)\n",
    "            x = np.array([[68/50, 68/50, 0], [63/50, 0, 63/50], [0, 0, 63/50],[0,54/50,54/50]])\n",
    "            x_t=x.transpose()\n",
    "            y=[dict_exons[exons_seq[0]],dict_exons[exons_seq[1]],dict_exons[exons_seq[2]],dict_exons[exons_seq[3]]]\n",
    "            x_t_x=np.dot(x_t,x)\n",
    "            x_t_x_minues = np.linalg.inv(x_t_x) \n",
    "            x_then=np.dot(x_t_x_minues,x_t)\n",
    "            beta=np.dot(x_then,y)\n",
    "            freq_li=[beta[0]/sum(beta),beta[1]/sum(beta),beta[2]/sum(beta)]\n",
    "            ret=ret+[(isoforms_seq[0],freq_li[0])]\n",
    "            ret=ret+[(isoforms_seq[1],freq_li[1])]\n",
    "            ret=ret+[(isoforms_seq[2],freq_li[2])]\n",
    "        else:\n",
    "            freq=1/isonum\n",
    "            for seq in isoforms_seq:\n",
    "                ret=ret+[(seq,freq)]\n",
    "    return ret"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'caagctagacctttctaggctcaaggacaaggctcgggtagcccagcttgtcgagttgtattgtccag': 767, 'tcgctaggacgcgtggggatgatatagcgcataccgaaggggtggtcttggtctggggccttt': 570, 'tgtatgactacatgccttggggatggtgttgttccactgctagatcgatgttccgcagtcggg': 883, 'ccaatctttcactccgattccggtccgtcgattgaaccactcgcctacaggcga': 914}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[('cccgtcagtgtgtgagcgtgtactcgctatatttgtagtgtgcactgcagaccaagtcggatgtttccttcccgaatctgtgtacacaaaacgtgagatctctcgatgtcacctggcaggtctgggtaagcgtctatgacttttggtgagttgagtccctacgatttgtttctattaacgctgttccatactgacaatgctggtacctaaccgctatccgggttattgtccgtgaggaatgagcagtattcccaaacgtgtcctcacgtgctgtg',\n",
       "  0.5),\n",
       " ('gtacacaaaacgtgagatctctcgatgtcacctggcaggtctgggtaagcgtctatgacttttggtgagttgagtccctacgatttgtttctattaacgctgttccatactgacaatgctggtacctaaccgctatccgggttattgtccgtgaggaatgagcagtattcccaaacgtgtcctcacgtgctgtg',\n",
       "  0.5),\n",
       " ('ccagtagggacgtatttaccggtggaagctggccgagattgtgcgtcacgaacgataaaaatggacaccccgtcgatgtaaggctgtcaaatttgggcgatcggttgctaacaacgcggcatactatatgctttgtgtatgttacgaagtagaggacctcatccaagattgccgccccaacgagtatttcgggaactcctggcaggccggttcgctcgcatatgtactacggccatgttacgggtat',\n",
       "  1.0),\n",
       " ('atcgggcgttaatagatatccaatgatagaccgacggccggggcactgatgtagcaagccgagatgggaatcgctgtatagtgtcaactgtatggaaagtgagtcggggtcagtttgatttatatgagtacagcatttcttgctcgtacctgct',\n",
       "  0.5),\n",
       " ('atcgggcgttaatagatatccaatgatagaccgacggccggggcactgatgtagcaagccgagatgggaatcgctgtatagtgtcaactgtatggaaagtgagtcggggtcagtttgatttatatgagtacagcatttcttgctcgtacctgcttaggtgggttagccctaggttatcatttataatttagtgctctccgatctcgtttgtttagaatgacctatgctagca',\n",
       "  0.5),\n",
       " ('caagctagacctttctaggctcaaggacaaggctcgggtagcccagcttgtcgagttgtattgtccagtcgctaggacgcgtggggatgatatagcgcataccgaaggggtggtcttggtctggggccttt',\n",
       "  0.02868432131887079),\n",
       " ('caagctagacctttctaggctcaaggacaaggctcgggtagcccagcttgtcgagttgtattgtccagccaatctttcactccgattccggtccgtcgattgaaccactcgcctacaggcga',\n",
       "  0.4556443331519144),\n",
       " ('tcgctaggacgcgtggggatgatatagcgcataccgaaggggtggtcttggtctggggccttttgtatgactacatgccttggggatggtgttgttccactgctagatcgatgttccgcagtcgggccaatctttcactccgattccggtccgtcgattgaaccactcgcctacaggcga',\n",
       "  0.5156713455292149),\n",
       " ('acgagcaatatcacctgaccgtttgacggggtacgcaaggtcgtgtctagttggaggctaagcgttacaacgttagcatggtctgatctaggagcctattccgccagaggtgcgtgcgactcattgcgacgaaacctcaattaaggttattgaattttggacatcagaaattgaaggtataggggca',\n",
       "  1.0)]"
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = parse_annotation_file('DATA_PA_1100_0.txt')\n",
    "genome = parse_genome_file('full_genome.txt')\n",
    "reads = parse_reads_file('shuffled_reads.txt')\n",
    "quantify_isoforms(genes, genome, reads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
